firebase-firestore/src/androidTest/java/com/google/firebase/firestore/TransactionTest.java

```diff
@@ -480,18 +473,50 @@ public void testCannotReadAfterWriting() {
     assertNotNull(e);
   }
 
+  @Test
+  public void testReadandUpdateNonExistentDocumentWithExternalWrite() {
+    final FirebaseFirestore firestore = testFirestore();
```

Comment:
Why `final` for one and not the other local that are referenced in the lambda below?

Note that Java 8 allows for references to variables from the lambda that are "effectively final" without requiring them to be actually final (which is why `counter` works as is). Consider just dropping the `final` here if that works.